src/test/kontrol/TermAuctionListInvariants.t.sol

```diff
@@ -0,0 +1,502 @@
+     * follow the assumption that unlockOffers will not revert.
+     */
+    function _guaranteeUnlockAlwaysSucceeds() internal {
+        bytes32 current = _termAuctionList.head;
+
+        while (current != TermAuctionList.NULL_NODE) {
+            PendingOffer storage offer = _termAuctionList.offers[current];
+            TermAuctionOfferLocker offerLocker =
+                TermAuctionOfferLocker(address(offer.offerLocker));
+
+            offerLocker.guaranteeUnlockAlwaysSucceeds();
+
+            current = _termAuctionList.nodes[current].next;
+        }
+    }
+
+    /**
+     * Assume that all RepoTokens in the PendingOffers have no discount rate
+     * set in the RepoTokenList.
+     */
+    function _assumeNoDiscountRatesSet() internal {
+        bytes32 current = _termAuctionList.head;
+
+        while (current != TermAuctionList.NULL_NODE) {
+            address repoToken = _termAuctionList.offers[current].repoToken;
+            uint256 discountRate = _repoTokenList.discountRates[repoToken];
+            vm.assume(discountRate == RepoTokenList.INVALID_AUCTION_RATE);
+
+            current = _termAuctionList.nodes[current].next;
+        }
+    }
+
+    /**
+     * Assume that all RepoTokens in the PendingOffers pass the checks performed
+     * in validateRepoToken, to ensure the function won't revert if they need to
+     * be inserted in the RepoTokenList.
+     */
+    function _assumeRepoTokensValidate(address asset) internal {
+        bytes32 current = _termAuctionList.head;
+
+        while (current != TermAuctionList.NULL_NODE) {
+            address repoToken = _termAuctionList.offers[current].repoToken;
+            (
+             uint256 redemptionTimestamp,
+             address purchaseToken,
+             ,
+             address collateralManager
+            ) = ITermRepoToken(repoToken).config();
+
+            vm.assume(purchaseToken != asset);
+            vm.assume(redemptionTimestamp < block.timestamp);
+
+            uint256 numTokens = ITermRepoCollateralManager(collateralManager).numOfAcceptedCollateralTokens();
+
+            for (uint256 i; i < numTokens; i++) {
+                address currentToken = ITermRepoCollateralManager(collateralManager).collateralTokens(i);
+                uint256 minCollateralRatio = freshUInt256();
+                _repoTokenList.collateralTokenParams[currentToken] = minCollateralRatio;
+
+                vm.assume(minCollateralRatio != 0);
+                vm.assume(
+                    ITermRepoCollateralManager(collateralManager).maintenanceCollateralRatios(currentToken) < minCollateralRatio
+                );
+            }
+
+            current = _termAuctionList.nodes[current].next;
+        }
+    }
+
+    /**
+     * Test that removeCompleted preserves the list invariants.
+     */
+    function testRemoveCompleted(address asset) external {
+        // For simplicity, assume that the RepoTokenList is empty
+        _repoTokenList.head = RepoTokenList.NULL_NODE;
```

Comment:
Since these tests focus on the invariants of the `TermAuctionList`, we are using an empty `RepoTokenList` here for simplicity, instead of initializing a general list of arbitrary size (like we do for the `TermAuctionList` in the `setUp()` function). The `removeCompleted` function may insert a token into this list, but the initial state of the `RepoTokenList` shouldn't otherwise affect the execution of the function or the `TermAuctionList` invariants. Note that the `insertSorted` function of `RepoTokenList` is already verified in `RepoTokenListInvariants.t.sol`.